app/src/main/java/com/voltix/wallet/service/CryptoPriceService.kt

```
@@ -0,0 +1,92 @@
+import com.voltix.wallet.common.SettingsCurrency
+import com.voltix.wallet.data.common.data_store.AppDataStore
+import com.voltix.wallet.models.Coin
+import com.voltix.wallet.models.Coins
+import com.voltix.wallet.models.CryptoPrice
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.async
+import kotlinx.coroutines.withContext
+import java.io.BufferedReader
+import java.io.InputStreamReader
+import java.net.URL
+import java.util.Date
+import javax.inject.Inject
+
+class CryptoPriceService @Inject constructor(
+    private val appDataStore: AppDataStore
+) {
+    private val cache: MutableMap<String, Pair<CryptoPrice, Date>> = mutableMapOf()
+
+    suspend fun getPrice(priceProviderId: String): Double {
+        var price = 0.0
+
+        withContext(Dispatchers.IO) {
+            val priceCoinGecko = async { getAllCryptoPricesCoinGecko() }.await()
+
+            if (priceCoinGecko != null) {
+                price = priceCoinGecko
+                    .prices[priceProviderId]
+                    ?.get(SettingsCurrency.getCurrency(appDataStore).name) ?: 0.0
+            }
+        }
+
+        return price
+    }
+
+    private suspend fun getAllCryptoPricesCoinGecko(): CryptoPrice? {
+        val coins = getCoins().joinToString(separator = ",")
+        return fetchAllCryptoPricesCoinGecko(coins, "USD")
+    }
+
+    private suspend fun fetchAllCryptoPricesCoinGecko(
+        coin: String = "bitcoin",
+        fiat: String = "usd"
+    ): CryptoPrice? {
+        val cacheKey = "$coin-$fiat"
+
+        if (cache.containsKey(cacheKey) && isCacheValid(cacheKey)) {
+            return cache[cacheKey]?.first
+        }
+
+        return try {
+            val response = fetchPrices(coin, fiat)
+            val decodedData = Gson().fromJson(response.toString(), CryptoPrice::class.java)
+            cache[cacheKey] = Pair(decodedData, Date())
+            decodedData
+        } catch (e: Exception) {
+            null
```

can we please log the exception ? don't swallow exceptions